src/main/java/com/playmonumenta/networkrelay/RemotePlayerManagerBungee.java

```diff
@@ -0,0 +1,216 @@
+		} catch (Exception e) {
+			MMLog.severe(() -> "Failed to get shard name");
+		}
+		if (shardName == null) {
+			throw new RuntimeException("Got null shard name");
+		}
+		return shardName;
+	}
+
+	static RemotePlayerBungee fromLocal(ProxiedPlayer player, boolean isOnline) {
+		@Nullable String targetShard = player.getServer() != null ? player.getServer().getInfo().getName() : ""; // TODO: add a way to get the shard name. IT IS MUCH EASIER IN VELOCITY - usb
+		return fromLocal(player, isOnline, targetShard);
+	}
+
+	static RemotePlayerBungee fromLocal(ProxiedPlayer player, boolean isOnline, String targetShard) {
+		return new RemotePlayerBungee(
+			getServerId(),
+			player.getUniqueId(),
+			player.getName(),
+			isOnline,
+			null,
+			targetShard
+		);
+	}
+
+	void refreshLocalPlayers() {
+		for (ProxiedPlayer player : ProxyServer.getInstance().getPlayers()) {
+			refreshLocalPlayer(player);
+		}
+	}
+
+	// Run this on local players whenever their information is out of date
+	void refreshLocalPlayer(ProxiedPlayer player) {
+		MMLog.fine(() -> "Refreshing local player " + player.getName());
+		RemotePlayerBungee localPlayer = fromLocal(player, true);
+
+		// update local player with data
+		if (updatePlayerLocal(localPlayer, false)) {
+			localPlayer.broadcast();
+		}
+	}
+
+	void remotePlayerChange(JsonObject data) {
+		if (data == null) {
+			MMLog.severe(() -> "Null player data recieved from an unknown source!");
+			return;
+		}
+		RemotePlayerAbstraction player = RemotePlayerAbstraction.from(data);
+		if (player == null) {
+			// something really bad happened
+			MMLog.severe(() -> "Invalid player data recieved from an unknown source!");
+			return;
+		}
+
+		updatePlayerLocal(player, true);
+	}
+
+	boolean updatePlayerLocal(RemotePlayerAbstraction player, boolean isRemote) {
+		RemotePlayerData oldPlayerData = getRemotePlayer(player.mUuid);
+		String serverType = player.getServerType();
+		RemotePlayerAbstraction oldPlayer = oldPlayerData != null ? oldPlayerData.get(serverType) : null;
+
+		// Update the player before calling events
+		super.updatePlayer(player);
+
+		if (player.mIsOnline && (oldPlayer == null || !oldPlayer.mIsOnline)) {
+			RemotePlayerLoadedEventBungee remotePE = new RemotePlayerLoadedEventBungee(player);
+			ProxyServer.getInstance().getPluginManager().callEvent(remotePE);
+			MMLog.info(() -> "Loaded player: " + player.mName + " remote=" + isRemote + " serverType=" + serverType);
+			return true;
+		}
+
+		@Nullable ProxiedPlayer localPlayer = ProxyServer.getInstance().getPlayer(player.mUuid);
+		if (isRemote && serverType.equals(RemotePlayerBungee.SERVER_TYPE) && localPlayer != null && localPlayer.isConnected()) {
+			// Player logged off on remote shard, but is locally online.
+			// This can happen if the remote shard was not notified the player logged in here in time.
+			MMLog.warning(() -> "Detected race condition, triggering refresh on " + player.mName + " remote=" + isRemote + " serverType=" + serverType);
```

Comment:
This is not severe enough to be a warning - it can happen fairly often as players switch shards and should resolve itself fast enough to be a non-issue.